hazelcast/src/hazelcast/client/network.cpp

```diff
@@ -700,163 +588,177 @@ namespace hazelcast {
+                    logger.warning("Exception during initial connection to ", address, ": ", e);
+                    return nullptr;
+                }
             }
 
-            void ClientConnectionManagerImpl::AuthCallback::scheduleTimeoutTask() {
-                auto executor = connection->getSocket().get_executor();
-                timeoutTimer = std::make_shared<boost::asio::steady_timer>(executor);
-                timeoutTimer->expires_from_now(connectionManager->connectionTimeoutMillis);
-                timeoutTimer->async_wait([=](boost::system::error_code ec) {
-                    if (ec) {
-                        return;
-                    }
-                    authFuture->onFailure(
-                            std::make_exception_ptr((exception::ExceptionBuilder<exception::TimeoutException>(
-                                    "ClientConnectionManagerImpl::authenticate")
-                                    << "Authentication response did not come back in "
-                                    << connectionManager->connectionTimeoutMillis.count() << " millis").build()));
-                });
+            void ClientConnectionManagerImpl::handleSuccessfulAuth(const std::shared_ptr<Connection> &connection,
+                                                                   auth_response response) {
+                check_partition_count(response.partition_count);
+                connection->setConnectedServerVersion(response.server_version);
+                connection->setRemoteAddress(std::move(response.address));
+                connection->setRemoteUuid(response.member_uuid);
+
+                auto new_cluster_id = response.cluster_id;
+                boost::uuids::uuid current_cluster_id = cluster_id_;
+
+                if (logger.isFinestEnabled()) {
+                    logger.finest("Checking the cluster: ", new_cluster_id, ", current cluster: ", current_cluster_id);
+                }
+
+                auto initial_connection = activeConnections.empty();
+                auto changedCluster = initial_connection && !current_cluster_id.is_nil() && !(new_cluster_id == current_cluster_id);
+                if (changedCluster) {
+                    logger.warning("Switching from current cluster: ", current_cluster_id, " to new cluster: ", new_cluster_id);
+                    client.getHazelcastClientImplementation()->on_cluster_restart();
+                }
+
+                activeConnections.put(response.member_uuid, connection);
+
+                if (initial_connection) {
+                    cluster_id_ = new_cluster_id;
+                    fire_life_cycle_event(LifecycleEvent::LifecycleState::CLIENT_CONNECTED);
+                }
+
+                logger.info("Authenticated with server ", response.address , ":", response.member_uuid
+                            , ", server version: " , response.server_version
+                            , ", local address: ", *connection->getLocalSocketAddress());
```

Comment:
What happens if you dereference an empty optional?

Reply:
if empty, dereferencing is either an illegal memory access or null ptr access, the process will die. I assumed that it could not be null at this point but it may be an oversight if the client disconnects immediately after successful authentication for some network failure. Hence, I am changing it to 
```                auto local_address = connection->getLocalSocketAddress();
                if (local_address) {
                    logger.info("Authenticated with server ", response.address , ":", response.member_uuid
                            , ", server version: " , response.server_version
                            , ", local address: ", *local_address);
                } else {
                    logger.info("Authenticated with server ", response.address , ":", response.member_uuid
                            , ", server version: " , response.server_version
                            , ", no local address (connection disconnected ?)");
                }
```